keps/0023-enable-disable-features.md

```diff
@@ -0,0 +1,131 @@
+---
+kep-number: 23
+title: Enabling and Disabling features in KUDO Operators
+authors:
+  - "@zmalik"
+owners:
+  - "@zmalik"
+creation-date: 2020-01-28
+last-updated: 2020-01-28
+status: provisional
+---
+
+# Enabling and Disabling features in KUDO Operators
+
+## Summary 
+
+KUDO enables Operator developers to create Operators and develop features for them. 
+The goal of this KEP is to facilitate operator developers to enable and disable the features in a deterministic way.
+
+## Motivation
+
+We want to provide operator users a way to enable and disable features in their operator. And right now there is no operator-user friendly or operator-developer friendly way to do it. 
+
+### Goals
+
+Given a feature, `A` is composed of Kubernetes object `x`, `y` and `z`
+ 
+- create or update kubernetes objects  `x`, `y` and `z`, when feature `A` is enabled.
+- delete kubernetes objects  `x`, `y` and `z`, when feature `A` is disabled.
+
+The goal of this KEP is to provide a way to create and delete objects `x`, `y` and `z`. 
+
+### Non-Goals
+
+- complex logic like loops to trigger multiple features
+- enable/disable nested features
+- evaluate multiple conditions to enable one feature
+
+## Proposal
+
+### Using Instance Parameters
+
+In KUDO, specific plans can be triggered when updating a parameter. Plans are specifically formed by phases, steps and utimately by tasks.
+And when talking about enabling/disabling features, we are eventually talking about the `Kind` of the Task. Currently, we have `Apply`, `Delete` and `Pipe` tasks kind.
+Enabling a feature is running a task of `Apply` kind and disabling is running a task of `Delete`
+
+We propose introducing a new Task kind named `Feature`. The Feature kind handles Apply or Delete resources based on the value of a parameter in the Operator Instance.
+
+A `Feature` task will look like:
+
+```
+tasks:
+  - name: mirror-maker-configuration
+    kind: Feature
+    spec:
+      parameter: MIRROR_MAKER_ENABLED
+      resources:
+        - mirror-maker-cm.yaml
+  - name: mirror-maker
+    kind: Feature
+    spec:
+      parameter: MIRROR_MAKER_ENABLED
+      resources:
+        - mirror-maker.yaml
+```
+
+
+the parameter used in task would be defined in the parameters for the Operator, and must evaluate to a boolean value.
+```
+apiVersion: kudo.dev/v1beta1
+parameters:
+  - name: MIRROR_MAKER_ENABLED
+    default: "false"
```

Comment:
as part of this we should add a linting rule that this could be only true/false? Too bad we don't have types here, because then anyone can come and change the value of this param to "whatever" and I am not sure what we do then...

Reply:
linting would help with default values maybe, but when somebody comes and update with a random value that cannot be directly mapped to a boolean value, we should fail the plan.

for when we have types, if we decide to do that, it would be easier as the error could be direct to parameter type related. And not a plan failing.

Reply:
> change the value of this param to "whatever" and I am not sure what we do then...

that would be a `FATAL_ERROR` 🤷‍♀

Reply:
@zmalik what do you think about having an implicit param for a toggle task?   why force a number of tasks on operator dev that "must" be followed in a specific pattern?

we could have an implicit param based on the task name for `mirror-maker-configuration`... which automatically follows the "rules" which also doesn't need to be linted because it is correct by convention.

I propose that `mirror-maker-configuration` or `mirror-maker-configuration-enabled` based on the task name.  To enable or disable is to pass `-p mirror-maker-configuration=false` to install/update with a default being true.

Reply:
regarding **default** perhaps the default is defined in the Toggle task.  Default `enabled` | `disabled` or `Enabled: true|false`

Reply:
I'd prefer it to be explicit. 
- It keeps everything you can define with `-p` in the params.yaml
- It gives the operator developer more freedom with regards to naming their parameters. (Which might not be a good thing ;)
- It allows one to use one param on multiple toggle-tasks.

If we go with implicits, there are a couple questions:
- Is this implicit variable allowed to be used somewhere else in templates?
- We probably would need validation to prevent a param with the same name to be defined in `params.yaml`